Update `join_keys` vignette (#208)
insightsengineering/teal.data: vignettes/join-keys.Rmd

  %\VignetteEncoding{UTF-8}
---

> ```{r, include=FALSE}
> devtools::load_all()
> ```

Reviewer: ```suggestion
```

## Overview

`teal.data` gives the possibility to define primary keys of a dataset and to define the relations to other datasets.

Each dataset can have a set of keys that inform its structure and relation to other tables by specifying:

- Columns consisting of the primary key
- Merge keys, a concept similar to `SQL`'s foreign key.

> Usually, an application developer needs to specify the keys manually, but in case of datasets named according to the `ADaM` standard, `teal` can assign the keys on its own.

Reviewer: This is not true and vignette `including-adam-data-in-teal` shouldn't be mentioned. CDISC keys are handled in this package by relevant default (default_cdisc_join_keys and cdisc_data)

Author: I was considering changing the documentation on that vignette, but that was wrong as the `default_cdisc_join_keys` lives here.

I've added a new section at the bottom. Please take a look at it.

Refer to `vignette("including-adam-data-in-teal", package = "teal")` for more information.

### Anatomy of `join_keys`

A new `join_keys` can be created by defining a set of primary and foreign keys _(it can also be created empty and then modified)_.
The `join_keys` object can be further extended by defining a parent-child relationship between datasets, as well as adding/modifying/removing keys.

The `join_keys` function is used to specify keys:

- `join_keys(...)` is a collection of multiple `join_key` entries
- `join_key(...)` specifies the relation between two datasets:
  - `dataset_1`, `dataset_2` - name of each datasets (if `dataset_2` is the same as `dataset_1` then it creates a primary key)
  - `key` - (optionally) named vector of column names

> Note that it is assumed that join keys are symmetric, i.e. `join_key("x", "y", "x_col" = "y_col")` will enable merge
> from "x" to "y" and vice versa.

Reviewer: ```suggestion
Note that it is assumed that join keys are symmetric, i.e. `join_key("x", "y", c("x_col" = "y_col"))` will set relationship from "x" to "y" and vice versa.
```

| `##   --> ds1: [id1]`          | arrow `-->` denotes `ds2` is a child of `ds1`                                                                              | 
| `##   --* (implicit via parent with): ds3` |  **Implicit relationship between `ds2` & `ds3`**<br />_(given that they share common keys with same parent)_    |
| `## ds3: [id1, id3]`           |                                                                               | 
| `##   --> ds1: [id1]`          |                                                                               | 
| `##   --* (implicit via parent with): ds2` |                                                                               | 
| `## ds4: [no primary keys]`    |                                                                               | 
| `##   <-> ds5: [id5]`          |  **Foreign keys** <br />_(arrow `<->` denotes no parent definition between datasets)_      | 
| `## ds5: [no primary keys]`    |                                                                               | 
| `##   <-> ds4: [id4] `         |                                                                               | 

> 

Reviewer 2: I love the idea of a table like this. The only improvement would be row borders so it's easy to read.
<img width="856" alt="Screenshot 2023-11-28 at 12 58 22 PM" src="https://github.com/insightsengineering/teal.data/assets/49812166/e2e7b2fe-5689-401c-990b-52f902d4988b">

Author: I think the template being used on the site will add those rows

From [another vignete](https://insightsengineering.github.io/teal.data/latest-tag/articles/using-delayed-data-basic.html)

![image](https://github.com/insightsengineering/teal.data/assets/211358/bee46ad8-3413-4c75-8de3-b67c3825db25)

### Primary key

The simplest method to define the join keys when calling the `teal_data` function is to use the `join_keys` argument.
We can specify the column(s) of the dataset that _(together)_ uniquely identify rows in the dataset.

```{r, message=FALSE}
library(teal.data)
library(dplyr)

iris_with_keys <- mutate(iris, id = factor(row_number()))

td <- teal_data(
  IRIS = iris_with_keys,
  code = "IRIS <- iris %>% mutate(id = factor(row_number()))",
  join_keys = join_keys(join_key("IRIS", keys = "id"))

  ## Alternatively, when only a single primary key is defined the `join_keys()`
  ##  wrapper can be skipped

>   # join_keys = join_key("IRIS", keys = "id")

Reviewer 2: ```suggestion
  # join_keys = join_key("IRIS", keys = "id") # nolint
```

## Anatomy of Join Keys

A new `join_keys` can be created by defining a set of primary and foreign keys. 
The `join_keys` object can also be extended by establishing parent-child relationships between datasets and adding/modifying/removing keys.

The `join_keys` function is used to specify keys:

- `join_keys(...)` is a collection of multiple `join_key` entries
- `join_key(...)` specifies the relation between two datasets:
  - `dataset_1`, `dataset_2`: names of the datasets (if `dataset_2` is the same as `dataset_1`, it creates a primary key)
  - `key` (optional): named vector of column names

Note that join keys are assumed to be symmetric, i.e., `join_key("x", "y", c("x_col" = "y_col"))` establishes a relationship from "x" to "y and vice versa.

```{r, results="hide", message=FALSE, tidy=FALSE}
library(teal.data)
jk <- join_keys(
  join_key("ds1", keys = "id1"),                  # ds1: [id1]
  join_key("ds2", keys = c("id1", "id2")),        # ds2: [id1, id2]
  join_key("ds3", keys = c("id1", "id3")),        # ds3: [id1, id3]
  join_key("ds1", "ds2", keys = "id1"),           # ds1 <--> ds2
  join_key("ds1", "ds3", keys = "id1"),           # ds1 <--> ds3

>   join_key("ds4", "ds5", keys = c("id4" = "id5")) # ds4 <--> ds5

Reviewer 3: unfortunately : (

>   join_key("ds4", "ds5", keys = c("id4" = "id5")) # ds4 <--> ds5

Author: I've added the `tidy=FALSE` on the chunk options that will ignore `styler` it for this chunk.

Having the comments aligned is informative, but I can remove that option and just eat up the extra spaces

```{r}
# Adding a new ds5 <-> ds1 key
jk["ds5", "ds1"] <- "a_key"

# Removing an existing key
jk["ds4", "ds5"] <- NULL
```

## Join Keys Relationships

There are 2 types of relationships encoded with `join_keys` described in the following sections. The _primary_ and _foreign_ keys are created explicitly using the `join_key` function.

Additionally, the `join_keys` object detects implicit relationships when two datasets share foreign keys to a parent dataset, but not between themselves. These implicit relations are available just like another foreign key and can be used to merge datasets, despite not being defined by the user.

### Primary Key with _teal_data_

When using the `teal_data` function, the simplest method to define the join keys  is to use the `join_keys` argument.
We can specify the column(s) of the dataset that _(together)_ uniquely identify rows in the dataset.

```{r, include=FALSE}
# nolint start: commented_code_linter.
```

```{r, message=FALSE}
library(teal.data)
library(dplyr)

> iris_with_keys <- mutate(iris, id = factor(row_number()))
> 
> td <- teal_data(
>   IRIS = iris_with_keys,
>   code = "IRIS <- iris %>% mutate(id = factor(row_number()))",
>   join_keys = join_keys(join_key("IRIS", keys = "id"))
> 
>   ## Alternatively, when only a single primary key is defined the `join_keys()`
>   ## wrapper can be skipped
>   # join_keys = join_key("IRIS", keys = "id")
> )
> 
> join_keys(td)

Reviewer: Couple comments about this code chunk:
- iris changes names twice `iris -> iris_with_keys -> IRIS`
- code is reproducible but doesn't follow what happens actually with the object and some might be confused
- we should discourage writing `teal_data(..., code)` and promote `within` instead. 
- Alternatively, we can skip the code because example focus should be on `join_keys`

```suggestion
library(teal.data)
library(dplyr)

iris_df <- mutate(iris, id = factor(row_number()))

td <- teal_data(
  iris_df = iris_df,
  join_keys = join_keys(join_key("IRIS", keys = "id"))
)

join_keys(td)
```

The same applies to the next example. Please pick one way of specifying data in this vignette.

Either
```r
library(teal.data)
library(dplyr)

iris_df <- mutate(iris, id = factor(row_number()))

td <- teal_data(
  iris_df = iris_df,
  join_keys = join_keys(join_key("IRIS", keys = "id"))
)

join_keys(td)
```

or
```r
library(teal.data)
library(dplyr)

td <- within(
  teal_data(),
  iris_df <- mutate(iris, id = factor(row_number()))
)

join_keys(td) <- join_keys(join_key("IRIS", keys = "id"))

join_keys(td)
```

Author: - Used `within` and added `datanames(...) <- c(...)` as the standard for the vignette.
  - It seems like promoting good practice

I've kept the examples using `teal_data` from the old vignette as it seemed like a useful and realistic way for the user to learn about `join_keys` _(the two are tightly related)_.

| `## ds4: [no primary keys]`    |                                                                               | 
| `##   <-> ds5: [id5]`          |  **Foreign keys** <br />_(arrow `<->` denotes no parent definition between datasets)_      | 
| `## ds5: [no primary keys]`    |                                                                               | 
| `##   <-> ds4: [id4] `         |                                                                               | 

### Primary key with `teal_data`

The simplest method to define the join keys when calling the `teal_data` function is to use the `join_keys` argument.
We can specify the column(s) of the dataset that _(together)_ uniquely identify rows in the dataset.

```{r, include=FALSE}
# nolint start: commented_code_linter.
```

```{r, message=FALSE}
library(teal.data)
library(dplyr)

iris_with_keys <- mutate(iris, id = factor(row_number()))

td <- teal_data(
  IRIS = iris_with_keys,
  code = "IRIS <- iris %>% mutate(id = factor(row_number()))",
  join_keys = join_keys(join_key("IRIS", keys = "id"))

  ## Alternatively, when only a single primary key is defined the `join_keys()`

>   ##  wrapper can be skipped

Reviewer 4: ```suggestion
  ## wrapper can be skipped
```

Author: I kinda like the second lines of comments to have the extra space, but that is arbitrary and personal.

Implemented it in 07982451. Thanks!

## Overview

The `teal.data` package provides a way to define primary keys for a dataset and establish relationships with other datasets.

Each dataset can be characterized by:

- Columns constituting the primary key
- Merge keys, analogous to `SQL` foreign keys

Typically, an application developer manually specifies these keys. However, for datasets following the `ADaM` standard, `teal.data` can automatically assign keys using the `default_cdisc_join_keys` object.

> Refer to the section ["`ADaM` `join_keys`"](#adam-join_keys) for details on using this object to select specific datasets.

Reviewer 5: I believe I tried it and it wasn't enough. It's been three days, though, I may remember wrong.

> Refer to the section ["`ADaM` `join_keys`"](#adam-join_keys) for details on using this object to select specific datasets.

Author: The link wasn't updated when section name was changed.

Corrected and "un-backticked" the section 5b04cdd9

##### Uses of _join_keys_ in _teal_ applications

The primary function of `join_keys` in `teal` applications is to facilitate the seamless [merging of datasets](https://insightsengineering.github.io/teal.transform/latest-tag/articles/data-merge.html) using `teal.transform`.

Additionally, it plays a role on the data filtering using the _[Filter Panel](https://insightsengineering.github.io/teal/main/articles/filter-panel.html)_ in a `teal` application.

> The filtering of a _(parent)_ dataset triggers additional background actions that will look for all datasets in which foreign keys target that _parent_ and then filter those accordingly.

Author: Yup

Done

```diff
diff --git a/vignettes/join-keys.Rmd b/vignettes/join-keys.Rmd
index d97117ef..a3e937b5 100644
--- a/vignettes/join-keys.Rmd
+++ b/vignettes/join-keys.Rmd
@@ -27,7 +27,7 @@ Refer to the section ["Joining Keys with `ADaM` Datasets"](#join-keys-in-adam-da
 The primary function of `join_keys` in `teal` applications is to facilitate the seamless [merging of datasets](https://insightsengineering.github.io/teal.transform/latest-tag/articles/data-merge.html) using `teal.transform`.
 
 Additionally, it plays a role on the data filtering using the _[Filter Panel](https://insightsengineering.github.io/teal/main/articles/filter-panel.html)_ in a `teal` application.
-The filtering of a _(parent)_ dataset triggers additional background actions that will look for all datasets in which foreign keys target that _parent_ and then filter those accordingly.
+The filters applied to a _(parent)_ dataset are also applied to their children.
 
 ## Anatomy of Join Keys
```

> ## Anatomy of Join Keys

Author: Proposal:

![image](https://github.com/insightsengineering/teal.data/assets/211358/cefd3e5f-5c10-4fca-80ae-9fe9e898755b)

> ## Anatomy of Join Keys

Reviewer 5: These are comments, not specific editing requests:

* _by `dataname` reference_ is not clear.

* A new `join_keys` can also be created empty or with only a primary key.

* The `join_keys` function is used as a constructor as well as to get or set join keys in `teal_data` objects.

* Returning itself is unclear. The function returns itself?

* _Is_ a joining key "represented by a pair of datasets and a character vector"? Or is a joining key a (named) character vector that we know to be relevant to a pair of datasets? The way I always think about it is a joining key is a column mapping between two datasets. Maybe we can improve this phrasing a little?

| `##   --* (implicit via parent with): ds3` |  **Implicit relationship between `ds2` & `ds3`**<br />_(given that they share common keys with same parent)_    |
| `## ds3: [col_1, col_3]`           |                                                                               | 
| `##   --> ds1: [col_1]`          |                                                                               | 
| `##   --* (implicit via parent with): ds2` |                                                                               | 
| `## ds4: [no primary keys]`    |                                                                               | 
| `##   <-> ds5: [col_5]`          |  **Foreign keys** <br />_(arrow `<->` denotes no parent definition between datasets)_      | 
| `## ds5: [no primary keys]`    |                                                                               | 
| `##   <-> ds4: [col_4] `         |                                                                               | 

## Accessing and Modifying keys

> The `join_keys` object behaves like a list and the preferred method to get a specific key to use the _subset_ operator with 2 indices. 

Reviewer: @averissimo this is still not addressed

> The `join_keys` object behaves like a list and the preferred method to get a specific key to use the _subset_ operator with 2 indices. 

Author: Done in 598e66af

```{r}
# Using the jk object defined on "Anatomy of Join Keys"
jk

# Getting primary key of "ds1"
jk["ds1", "ds1"]

# Getting foreign keys between "ds2" and "ds1"
jk["ds4", "ds5"]

# Note the symmetry on the keys
jk["ds5", "ds4"]
```

> Modifying or adding a key uses the same notation with the assignment operator `<-`. This will ensure symmetry as it will create both relationships. Assigning `NULL` value will delete the relation.

Author: Done in 98d0be50

```{r}
# Adding a new ds5 <-> ds1 key
jk["ds5", "ds1"] <- "a_column"

# Removing an existing key
jk["ds4", "ds5"] <- NULL
```

## Join Keys Relationships

There are 2 types of relationships encoded with `join_keys` described in the following sections. The _primary_ and _foreign_ keys are created explicitly using the `join_key` function.

Additionally, the `join_keys` object detects implicit relationships when two datasets share foreign keys to a parent dataset, but not between themselves. These implicit relations are available just like another foreign key and can be used to merge datasets, despite not being defined by the user.

### Primary Key with _teal_data_

When using the `teal_data` function, the simplest method to define the join keys  is to use the `join_keys` argument.
We can specify the column(s) of the dataset that _(together)_ uniquely identify rows in the dataset.

```{r, include=FALSE}
# nolint start: commented_code_linter.
```

```{r, message=FALSE}
library(teal.data)
library(dplyr)

> td_pk <- within(

Author: Done in fecf32f4

  teal_data(),
  ds1 <- mutate(iris, id = factor(row_number()))
)
datanames(td_pk) <- c("ds1")

join_keys(td_pk) <- join_keys(join_key("ds1", keys = "id"))

join_keys(td_pk)
```

We can extend the previous example and define primary keys for multiple datasets:

```{r, message=FALSE}

> library(teal.data)

Reviewer: I agree with @averissimo  it is nicer when each code-chunk has own library calls so they are "independent".

> library(teal.data)

Author: I went back and forth a lot with this one, both with the library and the object raw build on every chunk (vs. progressive manipulation of an object).

From an outside perspective I think that this indendency works better.

> library(teal.data)

Reviewer 5: I'm fine with either as long as it's applied consistently. If packages is attached in every block, objects should start from scratch as well.

Just a note, I believe it is customary to attach the package only once.

For users familiar with `SQL` database schema, these relationships are symmetric and not as strict as `SQL` foreign key relationships as `teal` does not validate whether the values inserted into foreign key columns are present in the parent table.

For example:

```{r}
library(teal.data)

td_fk <- within(
  teal_data(),
  {
    ds1 <- data.frame(X = 1:10, Y = 21:30, Z = 1:10)
    ds2 <- data.frame(W = 10:1, V = 5:14, M = rep(1:5, 2))
    ds3 <- data.frame(V = 5:14, N = 4)
  }
)
datanames(td_fk) <- c("ds1", "ds2", "ds3")

join_keys(td_fk) <- join_keys(
  # Primary keys
  join_key("ds1", keys = c("X")),
  join_key("ds2", keys = c("V", "W")),
  join_key("ds3", keys = c("V")),
  # Foreign keys
  join_key("ds2", "ds1", c("W" = "X")),
  join_key("ds3", "ds2", c("V" = "V"))
)

# The parent of ds2 and ds3 is ds1
parents(td_fk) <- list(ds2 = "ds1", ds3 = "ds2")

join_keys(td_fk)
```

### Implicit Relationships

This is a special inferred relationship from existing `join_keys` that does not need to be explicitly defined, however, as any other it can be overwritten.

> If two datasets share common _foreign_ keys to the same _parent_ dataset, then there is an implicit relationship between them which is seamlessly accessible via the `join_keys` getter.

Author: Done in d2cb4c95

```{r}
library(teal.data)

td <- within(
  teal_data(),
  {
    ds1 <- data.frame(X = 1:10, Y = 21:30, Z = 1:10)
    ds2 <- data.frame(W = 10:1, V = 5:14, M = rep(1:5, 2))
    ds3 <- data.frame(V = 5:14, N = 4)
    ds4 <- data.frame(V = 5:14, R = rnorm(10))
  }
)
datanames(td) <- c("ds1", "ds2", "ds3", "ds4")

join_keys(td) <- join_keys(
  # Primary keys
  join_key("ds1", keys = c("X")),
  join_key("ds2", keys = c("V", "W")),
  join_key("ds3", keys = c("V")),
  join_key("ds4", keys = c("V")),
  # Foreign keys
  join_key("ds2", "ds1", c("W" = "X")),
  join_key("ds3", "ds2", c("V" = "V")),
  join_key("ds4", "ds1", c("V" = "X"))
)

# The parent of ds2 and ds3 is ds1
parents(td) <- list(ds2 = "ds1", ds3 = "ds2", ds4 = "ds1")

join_keys(td)

join_keys(td)["ds2", "ds4"]
```

_Note_ that the definition above contains no `join_key` for `"ds2" <-> "ds4"`

## _ADaM_ Join Keys

When working with datasets named according to the _ADaM_ standard, `teal.data` offers `default_cdisc_join_keys` object that contains primary and foreign keys for standard datasets.

```{r}
names(default_cdisc_join_keys) |> sort()
```

The full set of join keys can be used, or a subset of those.

> Please note, that if the subsetted dataset has a parent, that will also be returned.

Reviewer: @averissimo

> Please note, that if the subsetted dataset has a parent, that will also be returned.

Author: Done in e0982d7f

```{r}
# Adding a new ds5 <-> ds1 key
jk["ds5", "ds1"] <- "a_column"

# Removing an existing key
jk["ds4", "ds5"] <- NULL
```

## Join Keys Relationships

There are 2 types of relationships encoded with `join_keys` described in the following sections. The _primary_ and _foreign_ keys are created explicitly using the `join_key` function.

Additionally, the `join_keys` object detects implicit relationships when two datasets share foreign keys to a parent dataset, but not between themselves. These implicit relations are available just like another foreign key and can be used to merge datasets, despite not being defined by the user.

### Primary Key with _teal_data_

When using the `teal_data` function, the simplest method to define the join keys  is to use the `join_keys` argument.
We can specify the column(s) of the dataset that _(together)_ uniquely identify rows in the dataset.

```{r, include=FALSE}
# nolint start: commented_code_linter.
```

```{r, message=FALSE}
library(teal.data)
library(dplyr)

td_pk <- within(
  teal_data(),
  ds1 <- mutate(iris, id = factor(row_number()))
)
datanames(td_pk) <- c("ds1")

join_keys(td_pk) <- join_keys(join_key("ds1", keys = "id"))

> ## Alternatively, when only a single primary key is defined the `join_keys()`
> ## the wrapper can be skipped
> # join_keys(td_pk) <- join_key("ds1", keys = "id")

Reviewer: I wouldn't necessarily write about this. Let's promote one way and don't give too much informations (alternatives).

```suggestion
```

data_2 <- data.frame(W = factor(10:1), V = factor(5:14), M = rep(1:5, 2))

data <- teal_data(
  D1 = data_1,
  D2 = data_2,
  join_keys = join_keys(
    join_key("D1", keys = c("X")),
    join_key("D2", keys = c("V", "W"))
    ## equivalent to using primary_key

>     # join_key("D2", "D2", keys = c("V", "W")),

Reviewer 2: ```suggestion
    # join_key("D2", "D2", keys = c("V", "W")), # nolint
```

Author: Thanks, added a silent no lint to avoid having that code in the vignette

![image](https://github.com/insightsengineering/teal.data/assets/211358/7e4aad8c-4bf9-4875-b656-f68a870e6b67)

Reviewer 2: Even better!

data_2 <- data.frame(W = factor(10:1), V = factor(5:14), M = rep(1:5, 2))

data <- teal_data(
  D1 = data_1,
  D2 = data_2,
  join_keys = join_keys(
    join_key("D1", keys = c("X")),
    join_key("D2", keys = c("V", "W"))
    ## equivalent to using primary_key
    # join_key("D2", "D2", keys = c("V", "W"))
  ),
  code = c(
    "D1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)",
    "D2 <- data.frame(W = factor(10:1), V = factor(5:14), M = rep(1:5, 2))"
  )
)

join_keys(data)
```

```{r, include=FALSE}
# nolint end: commented_code_linter.
```

### Foreign keys with `teal_data`

> When passing multiple datasets to the `teal_data` function, dataset relationship are set using `join_keys` and `join_key` and these are used to merge datasets together within `teal` apps.

Reviewer 4: ```suggestion
When passing multiple datasets to the `teal_data` function, dataset relationships are set using `join_keys` and `join_key` and these are used to merge datasets together within `teal` apps.
```

  D1 = data_1,
  D2 = data_2,
  D3 = data_3,
  join_keys = join_keys(
    join_key("D1", "D2", c("X" = "W")),
    join_key("D2", "D3", c("V" = "V"))
  ),
  code = c(
    "D1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)",
    "D2 <- data.frame(W = factor(10:1), V = factor(5:14), M = rep(1:5, 2))",
    "D3 <- data.frame(V = factor(5:14), T = 4)"
  )
)

join_keys(data)
```

### Implicit relationships

This is a special inferred relationship from existing `join_keys` that does not need to be explicitly defined.
If two datasets that are a child of the same parent _(and have common keys)_, then there is an implicit relationship between them.

These implicit relationships can be used to merge 2 datasets together, just as if they were defined manually.

```{r}
jk <- join_keys(
  join_key("ds1", keys = "id1"),
  join_key("ds2", keys = c("id1", "id2")),
  join_key("ds3", keys = c("id1", "id3")),
  join_key("ds1", "ds2", keys = "id1"),
  join_key("ds1", "ds3", keys = "id1")
)

# The parent of ds2 and ds3 is ds1
parents(jk) <- list(ds2 = "ds1", ds3 = "ds1")

jk["ds3", "ds2"]
```

_note_: the definition above contains no `join_key` for `"ds2" <-> "ds3"`

### `ADaM` `join_keys`

> When working with datasets named according to the `ADaM` standard, `teal.data` can assign the keys on its own using the `teal.data::default_cdisc_join_keys` that contains primary and foreign keys for standard datasets (`r paste("\u0060", names(default_cdisc_join_keys), "\u0060", collapse = ", ", sep = "")`).

Reviewer 4: ```suggestion
When working with datasets named according to the `ADaM` standard, `teal.data` can assign the keys on its own using the `teal.data::default_cdisc_join_keys` that contains primary and foreign keys for standard datasets as mentioned below.
```{r echo=FALSE}
names(default_cdisc_join_keys)
```
```

How about this for better visibility of datasets?

Author: It looks great! thanks, done in 07982451

The full set of join keys can be used, or a subset of those.
Please note, that if the subsetted dataset has a parent, that will also be returned.

```{r}
default_cdisc_join_keys["ADSL"]

default_cdisc_join_keys["ADTTE"]

default_cdisc_join_keys[c("ADSL", "ADTTE", "ADRS")]
```

#### The full set of `ADaM` `join_keys`

```{r}
default_cdisc_join_keys

> ```

Author: I've changed the order, but didn't add `echo=FALSE` as that will remove any context of what the user is seeing.

However, I don't think that moving the full `default_cdisc_join_keys` to top will have any benefit as it will decrease the value of the other examples.

I won't die on this though :sweat_smile:

> ```

Reviewer: Im fin with current state already. I had the biggest issue with data in the examples, it looks good now
